### ISSUE TITLE
Add option to specify `Subject`

### ISSUE THREAD

**Issue by embano1** [knative-sandbox/kn-plugin-event#32]
Could not find any related issue/discussion. It would be nice to set a `ce-subject` in build/send commands. If you think that's useful I can create a PR.

**Comment by cardil** [comment]
Makes sense, although [Subject is considered optional](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#subject). I'm thinking that maybe a general way of setting optional Cloud Event fields be better?!? Something similar to `--field` arg, like:

```bash
kn event build \
  --ce-field subject=Acme \
  --field email=a@b.c
```

WDYT?

**Comment by embano1** [comment]
> Makes sense, although [Subject is considered optional](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#subject). I'm thinking that maybe a general way of setting optional Cloud Event fields be better?!? Something similar to `--field` arg, like:

I think that makes a nice API and could later be combined with a `--ce-version` flag to indicate and validate against a specific CE version (breaking/non-breaking).

So do you think of migrating the existing CE metadata fields as well, i.e. not only optional fields in `--ce-field` for consistency?

**Comment by cardil** [comment]
I feel like the non-optional fields should stay as they are, as they are often used. Exposing them would be easier to use and easier for newcomers to get familiar with CE basics - they can just invoke `kn event build --help` to get help on common options.

**Comment by embano1** [comment]
I understand your rational but that would ultimately lead to an inconsistent API, especially for newcomers I think.

So I'm in favor of harmonizing them with consistent prefixes, provide defaults (e.g. for required fields), descriptions and examples in the help section.

E.g. one might want to add additional context attributes later, which could be passed with `--ce-attribute` so everything is consistent for envelop fields.

**Comment by cardil** [comment]
All those required fields have some defaults already. Calling:

```
$ docker run --rm -p 8080:8080 quay.io/openshift-knative/knative-eventing-sources-event-display &
$ kn event send --to-url http://localhost:8080
Jun 30 13:56:36.582   INFO Event (ID: c7665942-d711-4036-849e-e87956ab34d4) have been sent.
☁️  cloudevents.Event
Validation: valid
Context Attributes,
  specversion: 1.0
  type: dev.knative.cli.plugin.event.generic
  source: kn-event/v0.3.0-8-g7ac0651
  id: c7665942-d711-4036-849e-e87956ab34d4
  time: 2021-06-30T11:56:36.580565532Z
  datacontenttype: application/json
Data,
  {}
```

So, the type, source, id and time are all set, by default.

**Comment by embano1** [comment]
Yes, I know they exist :)

But consider the changes for optional envelop fields, we would end up with different fields for:

- required fields (e.g. `id`)
- optional fields (e.g. `--ce-subject`)
- payload fields (`--field`)

So my thought was keeping `--field` but harmonizing required/optional envelop fields with `--ce-` prefix for consistency.

**Comment by embano1** [comment]
Closing this as there is no consensus.

**Comment by cardil** [comment]
> Closing this as there is no consensus.

I think it's worth having this ticket open for a while longer, as this is a real issue. We just need to settle on an API, and I think we could get to it shortly.

/cc @rhuss

**Comment by rhuss** [comment]
tbh,  for known fields i'm woudl support typed fields over key=value options for these reasons:

* Better help messages that show the possible options
* Bash/Zsh completion easily possible
* Better validation on the option parsing level
* Easier to memorize
* The usual way how to work on a UI

However, I probably would even omit the `ce-` prefix as it sounds redundant.

**Comment by cardil** [comment]
I have the same opinion as @rhuss on this. I think we should just add `--subject` option.

I didn't added it initially because I just missed it, tbh. For me, type and source are well enough. But, subject is part of CE spec so it should be supported here as well.
